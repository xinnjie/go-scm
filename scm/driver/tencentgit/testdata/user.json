{
  "id": 176405,
  "email": "wx_1600e990b7cb469a97087b334c94acee@git.code.tencent.com",
  "name": "xinnjie",
  "username": "xinnjie",
  "web_url": "https://git.code.tencent.com/u/xinnjie",
  "current_sign_in_at": "2022-02-22T15:42:06+0000",
  "is_admin": false,
  "projects_limit": 9999,
  "skype": null,
  "linkedin": null,
  "twitter": null,
  "theme_id": 1,
  "bio": null,
  "state": "active",
  "color_scheme_id": 4,
  "website_url": null,
  "created_at": "2021-08-12T12:16:46+0000",
  "identities": [
    {
      "extern_uid": "o2tOe0ek5eer_Z8s0Y-lkZGi_C78",
      "provider": "wechat_open_wx308f65d74dca84e2"
    },
    {
      "extern_uid": "o2tOe0ek5eer_Z8s0Y-lkZGi_C78",
      "provider": "wechat_mp_wxae833cca19422851"
    }
  ],
  "private_token": "ABCDE",
  "can_create_project": true,
  "avatar_url": "https://git.code.tencent.com/uploads/user/avatar/176405/7e72ab97e19e430287c34c54af3bf7e1."
}