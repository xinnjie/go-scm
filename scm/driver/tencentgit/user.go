// Copyright 2017 Drone.IO Inc. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package tencentgit

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/jenkins-x/go-scm/scm"
)

type userService struct {
	client *wrapper
}

func (s *userService) CreateToken(context.Context, string, string) (*scm.UserToken, *scm.Response, error) {
	return nil, nil, scm.ErrNotSupported
}

func (s *userService) DeleteToken(context.Context, int64) (*scm.Response, error) {
	return nil, scm.ErrNotSupported
}

func (s *userService) Find(ctx context.Context) (*scm.User, *scm.Response, error) {
	out := new(user)
	res, err := s.client.do(ctx, "GET", "api/v3/user", nil, out)
	return convertUser(out), res, err
}

func (s *userService) FindLogin(ctx context.Context, login string) (*scm.User, *scm.Response, error) {
	var resp *scm.Response
	var err error
	firstRun := false
	opts := scm.ListOptions{
		Page: 1,
	}
	for !firstRun || (resp != nil && opts.Page <= resp.Page.Last) {
		out := []*user{}
		path := fmt.Sprintf("api/v3/users?search=%s&%s", login, encodeListOptions(opts))
		resp, err = s.client.do(ctx, "GET", path, nil, &out)
		if err != nil {
			return nil, nil, err
		}
		firstRun = true
		for _, u := range out {
			if strings.EqualFold(u.Username, login) {
				return convertUser(u), resp, err
			}
		}
		opts.Page++
	}
	return nil, resp, scm.ErrNotFound
}

// FindLoginByID returns the scm.User object for the specified user id
func (s *userService) FindLoginByID(ctx context.Context, id int) (*scm.User, error) {
	path := fmt.Sprintf("api/v3/users/%d", id)
	out := &user{}
	resp, err := s.client.do(ctx, "GET", path, nil, &out)
	if err != nil {
		return nil, err
	}
	if resp.Status == http.StatusOK {
		return convertUser(out), err
	}
	return nil, scm.ErrNotFound
}

func (s *userService) FindEmail(ctx context.Context) (string, *scm.Response, error) {
	user, res, err := s.Find(ctx)
	return user.Email, res, err
}

func (s *userService) ListInvitations(context.Context) ([]*scm.Invitation, *scm.Response, error) {
	return nil, nil, scm.ErrNotSupported
}

func (s *userService) AcceptInvitation(context.Context, int64) (*scm.Response, error) {
	return nil, scm.ErrNotSupported
}

type user struct {
	ID        int    `json:"id"`
	Username  string `json:"username"`
	Name      string `json:"name"`
	WebUrl    string `json:"web_url"`
	Avatar    string `json:"avatar_url"`
	Email     string `json:"email"`
	CreatedAt string `json:"created_at"`
	IsAdmin   bool   `json:"is_admin"`
}

func convertUser(from *user) *scm.User {
	createAt, _ := time.Parse(timeFormat, from.CreatedAt)
	return &scm.User{
		ID:      from.ID,
		Avatar:  from.Avatar,
		Email:   from.Email,
		Login:   from.Username,
		Name:    from.Name,
		Created: createAt,
		Link:    from.WebUrl,
		IsAdmin: from.IsAdmin,
	}
}

func convertUserList(users []*user) []scm.User {
	dst := []scm.User{}
	for _, src := range users {
		dst = append(dst, *convertUser(src))
	}
	return dst
}
