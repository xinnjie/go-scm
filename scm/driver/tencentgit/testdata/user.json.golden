{
  "ID": 176405,
  "Login": "xinnjie",
  "Name": "xinnjie",
  "Avatar": "https://git.code.tencent.com/uploads/user/avatar/176405/7e72ab97e19e430287c34c54af3bf7e1.",
  "Email": "wx_1600e990b7cb469a97087b334c94acee@git.code.tencent.com",
  "Link": "https://git.code.tencent.com/u/xinnjie",
  "Created": "2021-08-12T12:16:46-00:00"
}